### PR TITLE
OpenAPI / Swagger2 spec generation

### PR DESCRIPTION
**Pull request by mbohlool**
This is alpha version of OpenAPI spec generation. Generated "/swagger.json" file (accessible on api server) is a valid OpenAPI spec with some warnings that will be fixed in next versions of spec generation. Currently it is possible to generate a client using this spec though I did not test the clients.

reference: #13414

**Release note**:

``` release-note
Alpha support for OpenAPI (aka. Swagger 2.0) specification served on /swagger.json (enabled by default)
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30233)

### CODE REVIEW

File: pkg/genericapiserver/openapi/exceptions.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	PathOperationID{"/api/v1/configmaps", "listNamespacedConfigMap"}:   "listAllNamespacesConfigMap",
+	PathOperationID{"/api/v1/endpoints", "listNamespacedEndpoints"}:    "listAllNamespacesEndpoints",
+	PathOperationID{"/api/v1/events", "listNamespacedEvent"}:           "listAllNamespacesEvent",
+	PathOperationID{"/api/v1/limitranges", "listNamespacedLimitRange"}: "listAllNamespacesLimitRange",
+}
+
+// fixOperation tries to rename operation IDs to make OpenAPI spec valid. The best solution to repeated
+// ops is to fix it in the source. If that is not possible, the opFix map would let customize the renaming. If the ID
+// is not in the opFix map, the name will be concatenated with a string computed from the path. see inferNamePartFromPath
+// for more information.
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/exceptions.go]: How do we fix generation so this is not necessary?

File: pkg/genericapiserver/openapi/openapi.go [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+func RegisterOpenApiService(config *Config, containers *restful.Container) (err error) {
+	o := openApi{
+		config: config,
+	}
+	err = o.buildSwaggerSpec()
+	if err != nil {
+		return err
+	}
+	containers.ServeMux.HandleFunc("/swagger.json", func(w http.ResponseWriter, r *http.Request) {
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: Is this path picked arbitrarily?

File: pkg/genericapiserver/openapi/openapi.go [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
+		}
+		resp.WriteAsJson(o.swagger)
+	})
+	return nil
+}
+
+func (o *openApi) buildSwaggerSpec() (err error) {
+	if o.swagger != nil {
+		fmt.Errorf("Swagger spec is already built. Duplicate call to buildSwaggerSpec is not allowed.")
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: Shouldn't this be a return?

File: pkg/genericapiserver/openapi/openapi.go [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+	if err != nil {
+		return err
+	}
+	o.swagger = &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:     "2.0",
+			Definitions: definitions,
+			Paths:       &paths,
+			Info: &spec.Info{
+				InfoProps: spec.InfoProps{
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: Probably not appropriate - this info needs to be passed in.  We're going to have lots of these and not all will strictly be Kubernetes

File: pkg/genericapiserver/openapi/exceptions.go (hunk 1 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/exceptions.go]: It should be done in api_installer. We should pass unique Operation name for each operation. There were just too many of them. I think it is better to have swagger 2.0 working and mark it as alpha, then fix those.

File: pkg/genericapiserver/openapi/openapi.go (hunk 2 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: No, according to swagger specification: "By convention, the Swagger specification file is named swagger.json."

File: pkg/genericapiserver/openapi/openapi.go (hunk 3 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: Correct. Fixed.

File: pkg/genericapiserver/openapi/openapi.go (hunk 4 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: Fixed.

File: pkg/genericapiserver/openapi/exceptions.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/exceptions.go]: Pass this function in to the installer as a fixup func so the implementation in this package knows nothing about Kube.

File: pkg/genericapiserver/openapi/openapi.go [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+			}
+		}
+	}
+	return definitions, nil
+}
+
+func buildModel(old_model swagger.Model) (ret spec.Schema, err error) {
+	ret = spec.Schema{}
+	// model.SubTypes is not used in go-restful, ignoring.
+	ret.Description = old_model.Description
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: style: Always use camelCase for variables in Kube

File: pkg/genericapiserver/openapi/openapi.go [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+	"github.com/go-openapi/spec"
+)
+
+type Config struct {
+	SwaggerConfig *swagger.Config
+	// List of webservice's path prefixes to ignore
+	IgnorePrefixes []string
+}
+
+type openApi struct {
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: style: API is an initialism so it should be capitalized - `openAPI`, `RegisterOpenAPIService`

File: pkg/genericapiserver/openapi/openapi.go [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/emicklei/go-restful"
+	"github.com/emicklei/go-restful/swagger"
+	"github.com/go-openapi/spec"
+)
+
+type Config struct {
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: Godoc on all public types and methods, and godoc on all fields of public types following standard Godoc rules ( must start with the field name and form a complete sentence).

File: pkg/genericapiserver/openapi/openapi.go (hunk 5 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: Fixed.

File: pkg/genericapiserver/openapi/openapi.go (hunk 7 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: Fixed.

File: pkg/genericapiserver/openapi/openapi.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: Will be good to add that as a comment in code so that its clear to anyone reading the code.

File: pkg/genericapiserver/openapi/openapi.go [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+			Info:        o.config.Info,
+		},
+	}
+	return nil
+}
+
+func (o *openAPI) buildModels() (definitions spec.Definitions, err error) {
+	definitions = spec.Definitions{}
+	for _, decl := range swagger.NewSwaggerBuilder(*o.config.SwaggerConfig).ProduceAllDeclarations() {
+		for _, oldNamedModel := range decl.Models.List {
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: Why calling it oldNamedModel?

File: pkg/genericapiserver/openapi/openapi.go [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+			definitions[oldNamedModel.Name], err = buildModel(oldNamedModel.Model)
+			if err != nil {
+				return spec.Definitions{}, err
+			}
+		}
+	}
+	return definitions, nil
+}
+
+func buildModel(oldModel swagger.Model) (ret spec.Schema, err error) {
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: oldModel -> swaggerModel?

File: pkg/genericapiserver/openapi/openapi.go [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+	ret = spec.Schema{}
+	// model.SubTypes is not used in go-restful, ignoring.
+	ret.Description = oldModel.Description
+	ret.Required = oldModel.Required
+	ret.Discriminator = oldModel.Discriminator
+	ret.Properties = make(map[string]spec.Schema)
+	for _, oldProp := range oldModel.Properties.List {
+		if _, ok := ret.Properties[oldProp.Name]; ok {
+			return spec.Schema{}, fmt.Errorf("Duplicate property in old swagger spec.")
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: Do you really want to return in this case?
Or log a warning and continue?

File: pkg/genericapiserver/openapi/openapi.go [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+		}
+		ret.Properties[oldProp.Name], err = buildProperty(oldProp)
+		if err != nil {
+			return ret, err
+		}
+	}
+	return ret, nil
+}
+
+func buildProperty(oldProperty swagger.NamedModelProperty) (newProperty spec.Schema, err error) {
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: similar comment as above. I think calling them swaggerProperty and openAPIProperty is better than calling them old and new property.

File: pkg/genericapiserver/openapi/openapi.go [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+
+// TODO(mbohlool): Convert default value type to the type of parameter
+func getDefaultValue(str swagger.Special) interface{} {
+	if len(str) > 0 {
+		return str
+	}
+	return nil
+}
+
+func buildOpenAPIType(oldType *string, oldFormat string) ([]string, string, error) {
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: Why name this buildOpenAPIType, instead of just buildType like others (buildModel, buildProperty). 
Though, I personally liked this name better, am fine either way - just be consistent.

File: pkg/genericapiserver/openapi/openapi.go [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+			if o.config.FixPath != nil {
+				path = o.config.FixPath(path)
+			}
+			inPathCommonParamsMap, err := findCommonParameters(routes)
+			if err != nil {
+				return paths, err
+			}
+			pathItem, exists := paths.Paths[path]
+			if exists {
+				return paths, fmt.Errorf("Duplicate webservice route has been found for path: %v", path)
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: same. Is warning and continuing better than returning error?

File: pkg/genericapiserver/openapi/openapi.go [hunk 14]
```diff
@@ -0,0 +1,485 @@
...
+}
+
+func (o *openAPI) buildOperations(route restful.Route, inPathCommonParamsMap map[string]spec.Parameter) (ret *spec.Operation, err error) {
+	ret = &spec.Operation{
+		OperationProps: spec.OperationProps{
+			Description: route.Doc,
+			Consumes:    route.Consumes,
+			Produces:    route.Produces,
+			ID:          route.Operation,
+			// TODO(mbohlool): should we hardcode this here? should we not list http?
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: Where is this surfaced in our current 1.2 spec?
How does go-restful decide?
Or is this a new field in 2.0?

File: pkg/genericapiserver/openapi/openapi.go [hunk 15]
```diff
@@ -0,0 +1,485 @@
...
+			if err != nil {
+				return ret, err
+			}
+			ret.Parameters = append(ret.Parameters, newParam)
+		}
+	}
+	return ret, nil
+}
+
+func GroupRoutesByPath(routes []restful.Route) (ret map[string][]restful.Route) {
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: Why make this public?

File: pkg/genericapiserver/openapi/openapi.go [hunk 16]
```diff
@@ -0,0 +1,485 @@
...
+	default:
+		return ret, fmt.Errorf("Unknown restful operation kind : %v", restParam.Kind)
+	}
+	if !isSimpleDataType(restParam.DataType) {
+		return ret, fmt.Errorf("Restful DataType should be a simple type, but got : %v", restParam.DataType)
+	}
+	ret.Type = restParam.DataType
+	ret.Format = restParam.DataFormat
+	ret.UniqueItems = !restParam.AllowMultiple
+	// TODO(mbohlool): make sure they type of default value matches Type
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: they -> the

File: pkg/genericapiserver/openapi_fixes.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+import (
+	"bytes"
+	"github.com/go-openapi/spec"
+	"strings"
+	"unicode"
+)
+
+// Functions in this file customize generated openAPI spec to make it a valid spec.
+// In an ideal world, these function should not exists, however current implementation of routes
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi_fixes.go]: exists ->exist

File: pkg/genericapiserver/openapi_fixes.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+// (manly in api_installer.go) make it hard to fix all of these in a timely manner.
+
+type pathOperationID struct {
+	path string
+	id   string
+}
+
+var opFix = map[pathOperationID]string{
+	pathOperationID{"/api/v1/configmaps", "listNamespacedConfigMap"}:   "listAllNamespacesConfigMap",
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi_fixes.go]: Can you add a comment why each of these require fixing? (Feel free to point to existing issues)

File: pkg/genericapiserver/openapi/openapi.go [hunk 19]
```diff
@@ -0,0 +1,485 @@
...
+				case "PUT":
+					pathItem.Put = op
+				case "DELETE":
+					pathItem.Delete = op
+				case "OPTIONS":
+					pathItem.Options = op
+				case "PATCH":
+					pathItem.Patch = op
+				}
+				if o.config.FixOperation != nil {
```
Comment (nikhiljindal) [inline pkg/genericapiserver/openapi/openapi.go]: I was expecting FixOperation to be called in buildOperations, just like FixPath is called in buildPath.

**Comment by nikhiljindal** [comment]
Looks good, mostly naming comments.

Would be good to add unit tests for the new code.
We also have unit tests to test the swagger response from /api at server: https://github.com/kubernetes/kubernetes/blob/dfcb649ee8de87fda722a08e8488b746b667d305/pkg/master/master_test.go#L435.

We can add a simple test for swagger.json path.

**Comment by nikhiljindal** [comment]
I will suggest removing "fixes #13414" from description so that we dont close the issue once this PR is merged. While we will get swagger 2.0 spec with this PR, we still wont be able to use 2.0 specific features since we are converting 1.2 models to 2.0.

I like that
* The code in this PR (specifically `pkg/genericapiserver/openapi/openapi.go`) is not closely tied to kubernetes. I can use it with any genericapiserver. In fact, it should not be a lot of work to use it with any apiserver. In future, we can evolve it into a separate library like go-restful.
* Its not a lot of code. We are reusing models generated by go-restful which gives us a great start.

Thanks for working on this @mbohlool

**Comment by mbohlool** [comment]
You are right. I won't completely fixes #13414. I guess it is fine to add it as a reference for this PR. Thanks @nikhiljindal

File: pkg/genericapiserver/openapi/openapi.go (hunk 9 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: I avoided swaggerModel because the new model is still technically swagger (2.0). I would however like this naming if I add a comment to the top of file explaining swagger here means swagger 1.2 and not swagger 2.0 (that is first version OpenAPI).

File: pkg/genericapiserver/openapi/openapi.go (hunk 10 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: If there is a duplicate, I would prefer we fix it rather than ignoring it. The fact that it is not failing now means we do not have the duplicate.

File: pkg/genericapiserver/openapi/openapi.go (hunk 12 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: adding OpenAPI to all functions here seems to be redundant as the package name is openapi. I would remove it from this method to be consistent.

File: pkg/genericapiserver/openapi/openapi.go (hunk 15 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: It was a mistake, will fix it.

File: pkg/genericapiserver/openapi/openapi.go (hunk 19 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: buildOperations does not have the context to fix the operation. I can pass that context to it, but in my opinion, it makes the code more complex. I think it is ok to say we build the operation and the we fix it based on the context we have. Also this part of code should hopefully goes away after we fix api-installation code.

File: pkg/genericapiserver/openapi/openapi.go (hunk 14 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: This is a new field in 2.0.

**Comment by mbohlool** [comment]
@nikhiljindal @smarterclayton I decided to ditch the whole idea of having fixes. The main fix was OpID duplicates that I realized swagger will auto-generate OpID if it is empty. My library prefers go-restful operations to have unique IDs and will use them, but if they are not unique, it will fall back to use empty OpID and let swagger generate an Id based on path and method of the operation.

For the path fix, it is an special case with go-restful and does not have anything to do with kubernetes. go-restful fixes this in their swagger spec generation and I documented it and did the same.

For empty responses, I found it cleaner to accept a DefaultResponse and use "default" response code instead of 200. Default response code is defined in swagger and I prefer it to 200 (200 may not be always the case, even in successful operations).

Overall I found this new code much cleaner. I applied the rest of your comments if they are still applicable to this new design. Let me know if I forgot anything.

File: pkg/genericapiserver/openapi/openapi.go (hunk 19 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: FYI, I've removed fixOperation completely, see my other comment.

File: pkg/genericapiserver/genericapiserver.go [hunk 20]
```diff
@@ -892,6 +894,25 @@ func (s *GenericAPIServer) InstallSwaggerAPI() {
 		},
 	}
 	swagger.RegisterSwaggerService(swaggerConfig, s.HandlerContainer)
+	openAPIConfig := openapi.Config{
```
Comment (smarterclayton) [inline pkg/genericapiserver/genericapiserver.go]: This probably needs to be optional - I don't think we always want to include this, and this is not generic (in the way that apiserver is generic).  I think this needs to be in pkg/master.go, or have something that splits it out of this code path (where it can be invoked optionally.

File: pkg/genericapiserver/openapi/openapi.go [hunk 21]
```diff
@@ -0,0 +1,534 @@
...
+		}
+		ret.Properties[swaggerProp.Name], err = buildProperty(swaggerProp)
+		if err != nil {
+			return ret, err
+		}
+	}
+	return ret, nil
+}
+
+func buildProperty(swaggerProperty swagger.NamedModelProperty) (openAPIProperty spec.Schema, err error) {
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: Needs godoc describing the general pattern you're applying here

File: pkg/genericapiserver/openapi/openapi.go [hunk 22]
```diff
@@ -0,0 +1,534 @@
...
+			Swagger:     OpenAPIVersion,
+			Definitions: definitions,
+			Paths:       &paths,
+			Info:        o.config.Info,
+		},
+	}
+	return nil
+}
+
+func (o *openAPI) buildModels() (definitions spec.Definitions, err error) {
```
Comment (smarterclayton) [inline pkg/genericapiserver/openapi/openapi.go]: godoc on the general pattern being applied here

**Comment by smarterclayton** [comment]
Does look more straightforward, a few comments on my end.

File: pkg/genericapiserver/genericapiserver.go (hunk 20 above)
Comment (mbohlool) [inline pkg/genericapiserver/genericapiserver.go]: What is the difference between this and swagger 1.2? Both of them (together) are optional and controlled by a single flag Config.EnableSwaggerSupport that I think make sense.

File: pkg/genericapiserver/openapi/openapi.go (hunk 22 above)
Comment (mbohlool) [inline pkg/genericapiserver/openapi/openapi.go]: No entirely sure what you mean by that. I will add a godoc comment for these methods, though they are private and (I think somehow) self explanatory by their signature.

File: pkg/genericapiserver/genericapiserver.go (hunk 20 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: Re: this not being generic enough: Yes, spec.Info in particular I agree is not generic. For ex: we are planning to add a heapster server that will use genericapiserver. We want to expose swagger API there (and hence EnableSwaggerSupport will be true), but the title we want there will be different than "Kubernetes APIs". The version will be different too.
Same for openshift apiserver using genericapiserver library.

How about passing spec.Info as a different flag?

Re: having different flags for enabling 1.2 swagger spec and this 2.0 spec: I am fine keeping a single flag, but its not a lot of work to do if @smarterclayton or anyone feels strongly.

File: pkg/genericapiserver/genericapiserver.go [hunk 23]
```diff
@@ -892,6 +894,25 @@ func (s *GenericAPIServer) InstallSwaggerAPI() {
...
+		SwaggerConfig:  &swaggerConfig,
+		IgnorePrefixes: []string{"/swaggerapi"},
+		Info: &spec.Info{
+			InfoProps: spec.InfoProps{
+				Title:   "Kubernetes APIs",
+				Version: "1.4",
```
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: We wont be able to keep it up to date.
I am fine with just removing it for now.

File: .gitignore [hunk 24]
```diff
@@ -107,3 +107,6 @@ kubernetes.tar.gz
 /.make/
 # Just in time generated data in the source, should never be commited
 /test/e2e/generated/bindata.go
+
+# This file used by some vendor repos to store secret variables and should not be ignored
```
Comment (nikhiljindal) [inline .gitignore]: Would suggest mentioning the repo that does this.

**Comment by nikhiljindal** [comment]
Are you planning to add unit tests for methods in pkg/genericapiserver/openapi/openapi.go?

**Comment by mbohlool** [comment]
@nikhiljindal Do all private methods need unit test too?

File: pkg/genericapiserver/genericapiserver.go (hunk 20 above)
Comment (mbohlool) [inline pkg/genericapiserver/genericapiserver.go]: I understand now, @smarterclayton sorry for my misunderstanding. Yes the part of config that are specific to kubernetes need to be moved to genericapiserver's config and be set in master.

**Comment by nikhiljindal** [comment]
Would be great if we can have unit tests for all methods with non-trivial code.
I will leave it up to you to make a call on which functions in this PR need tests.

File: pkg/genericapiserver/genericapiserver.go (hunk 23 above)
Comment (mbohlool) [inline pkg/genericapiserver/genericapiserver.go]: version is required. I will set it to "unversioned"

**Comment by mbohlool** [comment]
I've added some test that covers most of the code. This is ready for another round of review.

File: cmd/kube-apiserver/app/server.go [hunk 25]
```diff
@@ -272,6 +272,7 @@ func Run(s *options.APIServer) error {
 	genericConfig.ProxyDialer = proxyDialerFn
 	genericConfig.ProxyTLSClientConfig = proxyTLSClientConfig
 	genericConfig.Serializer = api.Codecs
+	genericConfig.OpenAPIInfo.Title = "Kubernetes"
```
Comment (smarterclayton) [inline cmd/kube-apiserver/app/server.go]: Do you also have to set version here?

**Comment by smarterclayton** [comment]
Has verification failures

File: cmd/kube-apiserver/app/server.go (hunk 25 above)
Comment (mbohlool) [inline cmd/kube-apiserver/app/server.go]: newConfig set both Title and Version to "Generic API Server" and "unversioned". here we just overwrite the title.

**Comment by nikhiljindal** [comment]
Thanks, LGTM.
Feel free to self apply the label when you get all checks to pass.

**Comment by mbohlool** [comment]
@k8s-bot test this issue #30407

**Comment by mbohlool** [comment]
@k8s-bot test this issue #30407

File: Godeps/Godeps.json [hunk 26]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "k8s.io/kubernetes",
-	"GoVersion": "go1.6",
+	"GoVersion": "go1.7",
```
Comment (piosz) [inline Godeps/Godeps.json]: Was this change intentional? If not please revert.

cc @wojtek-t

Comment (mbohlool) [inline Godeps/Godeps.json]: no it was not. sent the PR to revert it.

**Comment by nikhiljindal** [comment]
Now that we have a separate param, we should mention in the release note that admins can turn off this new endpoint by setting EnableOpenAPISupport = false.

However for admins to be able to turn it off, we will need to expose it as a command line flag in server_run_options and use that to set config.EnableOpenAPISupport.

**Comment by smarterclayton** [comment]
Is there a downside to this endpoint being exposed?

**Comment by mbohlool** [comment]
@nikhiljindal, @smarterclayton  Sorry I missed these comment. I see swagger is enabled by default and we do not have a server config for it. I don't see any downside having this enabled by default, but I can add server config if you think that is useful.

**Comment by nikhiljindal** [comment]
I am fine with keeping this enabled by default - should probably mention that explicitly in release notes. Right now it says "Alpha support". We use alpha at other places to mean disabled by default.

Re: allowing admins to disable it. Its generally good to allow admins to turn off new features. Not critical.

**Comment by mbohlool** [comment]
Sure, I will send another PR to allow admins to disable it after my improvement #31468 got merged. Do we still have time to change release note of 1.4?
